Edgars Andersons

eandersons opened the pull request:
*FooBarQix* and *InfQixFoo* implementation by Edgars Andersons.

eandersons committed:
Docker support

eandersons committed:
Unit tests

eandersons committed:
Foo, Bar implementation

eandersons committed:
Minor cosmetic changes in `README.md`

Changes were made based on VSCode extenstion's `markdownlint` output.

eandersons committed:
Unit tests were updated to introduce Qix

Some existing test methods in `tests/FooBarQix.php` were updated to adjust
their name or documentation block for the expected changes.

eandersons committed:
Minor cosmetic changes in `step_2.md`

Changes were made based on VSCode extenstion's `markdownlint` output.

eandersons committed:
Qix implementation

eandersons committed:
Minor cosmetic changes in `step_3.md`

Changes were made based on VSCode extenstion's `markdownlint` output.

eandersons committed:
Unit tests were updated for Foo, Bar and Qix occurence implementation

eandersons committed:
`service/FooBarQix.php` was slightly refactored

`service/FooBarQix.php` was updated to:

- these properties were renamed:
  - `$_barMultiplier` to `$_bar`;
  - `$_fooMultiplier` to `$_foo`;
  - `$_qixMultiplier` to `$_qix`;
- a new property `$_valuesMap` was added to store string representation for
  `$_foo`, `$_bar` and `$_qix`;
- the constructor method was updated to initialise the new property
  `$_valuesMap`;
- the method `_processInput` was updated to use the new property `$_valuesMap`
  to generate output string to make it less hardcoded.

eandersons committed:
Foo, Bar and Qix occurrence implementation

Changes in `service/FooBarQix.php`:

- the method `_processInput` was renamed to `_checkMultipliers`;
- a new method `_checkOccurrences` was added to check if the input contains
  Foo, Bar or Qix;
- the constructor method was updated to call the new method
  `_checkOccurrences`.

Methods `testValueHasMultipliersAndContainsBarAndQix` and
`testInputCanBeString`  in `tests/FooBarQix.php` were updated to provide the
correct test values for assertions.

eandersons committed:
Method `testValueOnlyContainsFoo` was updated in `tests/FooBarQixTest.php`

Method `testValueOnlyContainsFoo` was updated in `tests/FooBarQixTest.php` to
provide correct test value for the assertion.

eandersons committed:
Minor cosmetic changes in `step_4.md`

Changes were mostly made based on VSCode extenstion's `markdownlint`
output.

eandersons committed:
Unit tests for the upcoming service InfQixFoo

eandersons committed:
InfQixFoo implementation

`service/FooBarQix.php` was refactored to use it as the base class for the new
service:

- these properties were made protected instead of private and renamed:
  - from `$_bar` to `$intA`;
  - from `$_foo` to `$intB`;
  - from `$_qix` to `$intC`;
- new properties that are string representation of multipliers and occurrences
  `$intA`, `$intB` and `$intC` were added:
  - `$strA`;
  - `$strB`;
  - `$strC`;
- new property `$separator` was added to be used to join together output parts;
- new static method `command` was added for processing input from the command
  line.

New class `InfQixFoo` (`service/InfQixFoo.php`) was added for the service
*InfQixFoo*.

`Instructions.md` was updated to include instructions for the service
*InfQixFoo*.

eandersons committed:
`tests/InfQixFooTest.php` was updated to add test for the upcoming feature

A new method `testSumOfAllValueDigitsIsDivisibleBy8` was added to the test
class `InfQixFooTest` to test if the sum of all value's digits is divisible by
8.

eandersons committed:
Minor cosmetic changes in `step_5.md`

Changes were made based on VSCode extenstion's `markdownlint` output.

eandersons committed:
New feature - check if sum of input's digits is divisible by 8

If sum of all input integer's digits is divisible by 8, `Inf` is appended to
the output.

Changes in the class `FooBarQix`:

- property `$_ouput` was changed from private to protected to make it
  accessible in subclasses and renamed to `$output`;
- constructor method was changed from private to protected to make it
  accessible in subclasses;
- new method `inputDigitsAsArray` was added to get an array that contains input
  integer's digits;
- method `_checkOccurrences` was updated to use the new method
  `inputDigitsAsArray` to get the necessary value.

The class `InfQixFoo` was updated to add the functionality to check if the sum
of input integer's digits is divisible by 8 (Inf) to add `Inf` to the output in
such cases.

`tests/InfQixFooTest.php` was updated to provide values in some tests that do
not clash with the new functionality.

eandersons committed:
Code coverage

`Dockerfile` was updated to install *Xdebug* that is necessary for *PHPUnit*'s
code coverage functionality.

Documentation block `FooBarQix::command` was updated to add the tag
`@codeCoverageIgnore` as this method is not directly related to the necessary
functionality.

`Instructions.md` was updated to include code coverage related commands.

OskarsOzols reviewed (changes requested):
This will be like general review of Your pull request.

There are three major problems we need You to fix in this code.

1) It's not scalable in terms of count of input params. Now You have variables like intA, intB, etc. You will soon run out of alphabet if there are more and from requirements You see that these type of changes could happen in the future.

2) InfQixFoo class is implemented based on Inheritance. Can You refactor it to be based on Composition?

3) Both implementation classes have a lot of mutability. Can You change them to become more (or max) immutable?

One minor test problem:

There are many duplicate tests (like testNegativeIntegerIsNotAValidInput) for both classes. We want to avoid duplication.

eandersons committed:
`FooBarQix` was updated to define transformation values in an array

These updates introduces flexibility as now indefinite amount of transformation
values may be provided.

Changes in `service/FooBarQix.php`:

- property `$_valuesMap` was renamed to `$changeMap` and made protected;
- method `_checkMultipliers` was completely rewritten to use the array with
  transformation values.

Changes in `service/InfQixFoo.php`:

- adjustments for the changes in the base class `FooBarQix`;
- method `_checkDivisibilityOfSumOfInputDigits` was updated to use the new
  properties `$changeMap` and `divisor`.

Composer files (`composer.json` and `composer.lock`) were updated to include
the new dependency `drupol/phpermutations` that is used in
`FooBarQix::_checkMultipliers`.

eandersons committed:
Unit tests for the upcoming class `Command`

eandersons committed:
Inheritance in the class `InfQixFoo` was replaced with composition

Service classes were updated to improve immutability.

New class `Command` was added for executing service commands. `Instructions.md`
was updated to reflect these changes.

eandersons committed:
Test classes were updated to avoid duplication

Changes in classes `FooBarQixTest` and `InfQixFooTest`:

- they now inherits from the class `CommonServiceTestCase`;
- these methods were moved to the new class `CommonServiceTestCase` to avoid
  duplication in classes `FooBarQixTest` and `CommonServiceTestCase`:
  - `testNegativeIntegerIsNotAValidInput`;
  - `testZeroIsNotAValidInput`;
  - `testFloatIsNotAValidInput`;
  - `testNonintegerStringIsNotAValidInput`;
  - `testValueOnlyHasMultiplierFoo`;
  - `testValueOnlyHasMultiplierQix`;
  - `testValueIsNotTransformed`.

`Composer.json` was updated to add tests directory to the class map.

eandersons commented:
@OskarsOzols, I have made changes that address problems you mentioned in the review.
